tapdb/multiverse.go

```diff
@@ -61,21 +70,238 @@ type BatchedMultiverse interface {
+
+	// First, get the sub-cache for this universe ID from the map of
+	// caches.
+	idStr := treeID(id.String())
+	proofCache, _ := p.LoadOrStore(idStr, newLeafCache())
+	assetProofCache := proofCache.(leafProofCache)
+
+	// With that lower level cache obtained, we can check to see if we have
+	// a hit or not.
+	proofKey := NewProofKey(id, leafKey)
+	proof, err := assetProofCache.Get(proofKey)
+	if err == nil {
+		log.Debugf("read proof for %v+%v from cache, key=%v",
+			id.StringForLog(), leafKey, proofKey[:])
+		return *proof
+	}
+
+	return nil
+}
+
+// insertProof inserts the given proof into the cache.
+func (p *proofCache) insertProof(id universe.Identifier, leafKey universe.LeafKey,
+	proof []*universe.Proof) {
+
+	idStr := treeID(id.String())
+
+	proofCache, _ := p.LoadOrStore(idStr, newLeafCache())
+	assetProofCache := proofCache.(leafProofCache)
+
+	proofKey := NewProofKey(id, leafKey)
+
+	log.Debugf("storing proof for %v+%v in cache, key=%x",
+		id.StringForLog(), leafKey, proofKey[:])
+
+	proofVal := cachedProof(proof)
+	assetProofCache.Put(proofKey, &proofVal)
+}
+
+// delProofsForAsset deletes all the proofs for the given asset.
+func (p *proofCache) delProofsForAsset(id universe.Identifier) {
+	log.Debugf("wiping proofs for %v from cache", id)
+
+	idStr := treeID(id.String())
+	p.Delete(idStr)
+}
+
+// cachedRoots is used to store the latest root for each known universe tree.
+type cachedRoots []universe.Root
+
+// atomicRootCache is an atomic pointer to a root cache.
+type atomicRootCache = atomic.Pointer[cachedRoots]
+
+// newAtomicRootCache creates a new atomic root cache.
+func newAtomicRootCache() atomicRootCache {
+	treeCache := &cachedRoots{}
+
+	var a atomicRootCache
+	a.Store(treeCache)
+
+	return a
+}
+
+// rootIndex maps a tree ID to a universe root.
+type rootIndex = lnutils.SyncMap[treeID, *universe.Root]
+
+// atomicRootIndex is an atomic pointer to a root index.
+type atomicRootIndex = atomic.Pointer[rootIndex]
+
+// newAtomicRootIndex creates a new atomic root index.
+func newAtomicRootIndex() atomicRootIndex {
+	var a atomicRootIndex
+	a.Store(&rootIndex{})
+
+	return a
+}
+
+// rootNodeCache is used to cache the set of active root nodes for the
+// multiverse tree.
+type rootNodeCache struct {
+	sync.RWMutex
+
+	rootIndex atomicRootIndex
+
+	allRoots atomicRootCache
+
+	// TODO(roasbeef): cache for issuance vs transfer roots?
+}
+
+// newRootNodeCache creates a new root node cache.
+func newRootNodeCache() *rootNodeCache {
+	return &rootNodeCache{
+		rootIndex: newAtomicRootIndex(),
+		allRoots:  newAtomicRootCache(),
+	}
+}
+
+// fetchRoots reads the cached roots for the given proof type. If the amounts
+// are needed, then we return nothing so we go to the database to fetch the
+// information.
+func (r *rootNodeCache) fetchRoots(withAmts, haveWriteLock bool,
+) []universe.Root {
+
+	// We don't cache roots with amounts, so if the caller wants the
+	// amounts, they'll go to disk.
+	if withAmts {
```

Comment:
Non-blocking for this PR, but I'm wondering if we should add the option to disallow this call on public universe instances, since it bypasses the cache?